[WIP] Add AWS Query protocol encoders support

igor-ramazanov opened the pull request.

igor-ramazanov committed:
AWS Query Protocol encoders WIP

igor-ramazanov commented:
Not ready for a full-review yet, there are some things which need to be addressed first:
1. The [FormData](https://github.com/disneystreaming/smithy4s/pull/700/files#diff-ce5d21b72a5a8d0fd8aa6502e0ea8cc5fdc7a3324aff2d2547e2dccfd7fe4e36) data structure probably needs a better representation.
2. It's not clear to me if [primitives/structures/lists in tests](https://github.com/disneystreaming/smithy4s/pull/700/files#diff-66d1e93a40b30a3f035acbc022f2b29b323c2cc6a9ed6ab334b21ec34047fbe4R42) need to be prefixed with the structure name, e.g. `x=1` or `Foo.x=42`.

igor-ramazanov commented:
3. Perhaps, would be good to copy tests from https://github.com/awslabs/smithy/tree/main/smithy-aws-protocol-tests/model/awsQuery

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,55 @@
+package smithy4s.aws.query
+
+final case class FormData(
```
That's not bad at all, but I'd rather we represented it as follows (which allows to re-use some types) 

```scala
import smithy4s.PayloadPath

sealed trait FormData {
   def render: String = ??? 
   def prepend(key: String) = ???
   def prepend(index: Int) = ??? 
} 
object FormData {
    final case class SimpleValue(str: String)  extends FormData
    final case class PathedValue(path: PayloadPath, value: String) extends FormData 
    final case class MultipleValues(values: Vector[PathedValue]) extends FormData 
} 
```

The there is no look-up in the encoding side of things, so no need to use a `Map`. You may need to add `def prepend` methods to the existing `PayloadPath` type.

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,55 @@
+    values: Map[List[String], String],
+    primitive: Option[String]
+) {
+  lazy val render: Option[String] =
```
this can be kept a `def`, ~and return an empty string when it holds a primitive~ and return the string value when it holds a primitive (ie, no need for `Option`)

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,55 @@
+      )
+  }
+
+  def childOf(parent: String): FormData =
```
Looking at the [spec](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html?highlight=query%20prot#collections), this doesn't need to exist : the id of the shape is never used in the encoding.

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,55 @@
+        Some(
+          values
+            .map { case (segments, value) =>
+              segments.reverse.mkString(".") + "=" + value
```
In theory you should never have to reverse, as you're always gonna be computing the fields first and prepending the name of the structure field or the union alternative

Baccata replied:
Also, you're gonna have to URL-encode both the key and value here. You can find an encoder in `smithy4s.internals.URIEncoderDecoder`, which is package private, so you'll have to write the package of this file as : 

``` 
package smithy4s
package aws.query
```

igor-ramazanov replied:
Done

Baccata commented on modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala:
```diff
@@ -0,0 +1,354 @@
+object AwsQueryCodecSpec extends SimpleIOSuite {
+
+  implicit class SchemaOps[A](schema: Schema[A]) {
+    def named(name: String) = schema.withId(ShapeId("default", name))
```
you should be able to remove this in these tests:  the name of the shape is never to be taken into consideration in this encoding.

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+    val codecs: Vector[AwsQueryCodec[S]] =
+      fields.map(field => fieldEncoder(field))
+
+    val structKey = getKey(hints, shapeId.name)
```
that is un-necessary (and the `childOf` call needs to be removed)

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+      dispatch
+        .projector(alt)(u)
+        .fold(FormData.empty)(a => compile(alt.instance)(a))
+        .childOf(key)
```
this call should be replaced by a call to `prepend` (see my comment [there](https://github.com/disneystreaming/smithy4s/pull/700/files#r1060512960))

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+      codecs
+        .map(codec => codec(s))
+        .reduce(_ combine _)
+        .childOf(structKey)
```
the call to `childOf` needs to be removed here

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+        }
+      })
+
+      (s: S) => encoder(field.get(s)).childOf(fieldKey)
```
the `childOf` call should be replaced by a call to `prepend` (see my comment [there](https://github.com/disneystreaming/smithy4s/pull/700/files#r1060512960))

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+            label: String,
+            instance: Schema[AA],
+            get: S => AA
+        ): AwsQueryCodec[AA] = (a: AA) => compile(instance)(a)
```
So the tricky part here is that you need to pre-compile the schema instead of doing on a per-request basis. Essentially you need to do : 

``` 
val compiled = compile(instance) 
(a: AA) => compiled(a)
```

it seems the same as what you wrote, semantically, but it'll performs much better at runtime.

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+            instance: Schema[AA],
+            get: S => Option[AA]
+        ): AwsQueryCodec[Option[AA]] = {
+          case Some(value) => compile(instance)(value)
```
[ditto](https://github.com/disneystreaming/smithy4s/pull/700/files#r1060526010)

igor-ramazanov replied:
Done

Baccata commented on modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala:
```diff
@@ -0,0 +1,354 @@
+      implicit val schema: Schema[FooBar] =
+        enumeration[FooBar](List(Foo, Bar)).x
+    }
+    val xmlFoo = "x=foo"
```
this test is wrong : it should be `val xmlFoo = "foo"` (same for `xmlBar`)

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+        .iterator(collection)
+        .zipWithIndex
+        .map { case (member, index) =>
+          val newKey = ((index + 1).toString :: key).reverse.mkString(".")
```
here you should just be doing : 

``` 
memberWriter(member).prepend(index) 
```

there's no need to reverse, nor to call `.mkString` prematurely. The `makeString` call will happen naturally when the `.render` is called on `FormData`

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+      tag: Primitive[P]
+  ): AwsQueryCodec[P] =
+    (p: P) => {
+      Primitive.stringWriter(tag, hints) match {
```
👍

Baccata commented on modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala:
```diff
@@ -0,0 +1,161 @@
+  ): AwsQueryCodec[B] = (b: B) => compile(schema)(refinement.from(b))
+
+  override def lazily[A](suspend: Lazy[Schema[A]]): AwsQueryCodec[A] = (a: A) =>
+    suspend.map(schema => compile(schema)).value(a)
```
So this one is tricky. For starters : [ditto](https://github.com/disneystreaming/smithy4s/pull/700/files#r1060526010) 

Then, unfortunately, it's likely not gonna work like you'd think (recursion is tricky). You can't rely on SAM-conversions here, you're gonna need need to be explicit : 

```
new AwsQueryCodec[A] {
   // we want the compilation to happen in a deferred manner, but only once. 
   lazy val underlying = compile(schema).value
   def apply(a: A) : FormData = underlying(a)  
} 
```

igor-ramazanov replied:
Done

igor-ramazanov committed:
Bump SBT 1.8.0 => 1.8.2

igor-ramazanov committed:
Address code review comments

1. `FormData` as ADT.
2. use `smithy4s.PayloadPath` to represent paths.
3. Update tests.
4. Code passes the tests (except `recursive` tests.

igor-ramazanov commented:
@Baccata Hey, thanks for the thorough review! I've addressed the comments.
Few things:
1. For some reason I have to manually strip the initial `.` for the `PayloadPath => String`: https://github.com/disneystreaming/smithy4s/pull/700/files#diff-ce5d21b72a5a8d0fd8aa6502e0ea8cc5fdc7a3324aff2d2547e2dccfd7fe4e36R38
2. Pay closer attention to the [AWS Query traits](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#supported-traits). For now, only the `XmlName` and `XmlFlattened` supported.
3. Fix the `recursive` test. I guess [this test from AWS Query spec](https://github.com/awslabs/smithy/blob/main/smithy-aws-protocol-tests/model/awsQuery/input.smithy#L287) should help me.
4. Copy tests from [the AWS spec](https://github.com/awslabs/smithy/tree/main/smithy-aws-protocol-tests/model/awsQuery)?

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,57 @@
+     * @todo Understand the root reason and have a better solution for a workaround removing the '.' prefix.
+     */
+    override def render: String =
+      URIEncoderDecoder.encode(path.toString.stripPrefix(".")) + "=" +
```
the `.` exist because this Path construct is used in json decoding and I wanted to make it look more like `jq`. Really you shouldn't rely on `.toString` there, but rather deconstruct the path and fold the segments, which would prevent instantiating a string only to strip it of one char.

igor-ramazanov replied:
Done, perhaps not ideal.

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    override def prepend(index: Int): FormData =
+      PathedValue(PayloadPath.fromString(index.toString), str)
```
Use `PayloadPath(index)` instead

igor-ramazanov replied:
Done

Baccata commented on modules/aws/src/smithy4s/query/FormData.scala:
```diff
@@ -0,0 +1,57 @@
+    override def render: String = URIEncoderDecoder.encode(str)
+
+    override def prepend(key: String): FormData = {
+      PathedValue(PayloadPath.fromString(key), str)
```
Use `PayloadPath(key)` instead

igor-ramazanov replied:
Done

Baccata commented on modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala:
```diff
@@ -0,0 +1,350 @@
+  }
+
+  test("recursion") {
+    ignore("Not ready yet")
```
the implementation of `lazily` looks good to me. I think we need to amend the datatype used in the test to carry more information, because the protocol is lossy and the current test wouldn't show much. 

If you had a required field, it will make it naturally easier to test :  

```scala 
case class  IntList(head: Int, tail: Option[IntList]) 

val input = IntList(1, Some(IntList(2, None)))
val expected = "head=1&tail.head=2" 
```

igor-ramazanov replied:
Worked well indeed. Thanks, done

Baccata commented:
> Pay closer attention to the [AWS Query traits](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#supported-traits). For now, only the XmlName and XmlFlattened supported. 

In the scope of this PR, this is absolutely fine. 

So I think what I'm gonna do is prep an `aws-query-protocol` protected feature branch that we could target in this PR and my own xml-based PRs. We'll make some progress there until we have the protocol tests wired correctly, and until some manual testing showcase that the aws signature algorithm works properly, at which point we'll perform a release. 

That way, I'll deem this PR ready to go when my latest feedback is addressed

kubukoz commented on modules/aws-kernel/src/smithy4s/aws/AwsProtocol.scala:
```diff
@@ -16,16 +16,15 @@
 private[aws] object AwsProtocol {
   val supportedProtocols: List[ShapeTag[_]] =
-    List(AwsJson1_0, AwsJson1_1)
+    List(AwsJson1_0, AwsJson1_1, AwsQuery)
```
**praise:** glad you found this :D

igor-ramazanov committed:
Code review: enable 'recursion' test

igor-ramazanov committed:
Code review: use the `PayloadPath.apply`, manually destructure the `PayloadPath` instead of `.toString`

igor-ramazanov commented:
@Baccata Sounds reasonable. Pushed commits addressing the last comments.

igor-ramazanov committed:
Fix the failing test

Baccata commented:
@igor-ramazanov, you'll need to run `headerCreateAll` in SBT to fix the current issue.

igor-ramazanov committed:
Add headers

igor-ramazanov commented:
@Baccata Pushed

Baccata commented:
awesome, I'm merging into [disneystreaming:feature/aws-query-protocol](https://github.com/disneystreaming/smithy4s/tree/feature/aws-query-protocol)

igor-ramazanov commented:
@Baccata Would be nice to open a draft/wip PR from the https://github.com/disneystreaming/smithy4s/tree/feature/aws-query-protocol?rgh-link-date=2023-01-12T10%3A30%3A44Z to the release branch to track the overall visibility of the progress.

Also, what are the next steps will be? I guess, wiring up all the stuff together and polishing the last details and corner cases?

Baccata commented:
> Also, what are the next steps will be? I guess, wiring up all the stuff together and polishing the last details and corner cases? 

Sorry I've been busy today, but I'll create some tickets asap for you to get a good idea of what's left and get to the point of a working client asap. 

> @Baccata Would be nice to open a draft/wip PR from the https://github.com/disneystreaming/smithy4s/tree/feature/aws-query-protocol?rgh-link-date=2023-01-12T10%3A30%3A44Z to the release branch to track the overall visibility of the progress.

Yup, I'm literally signing off now but I'll do that later or tomorrow. Thanks for the patience.
